lib/internal/streams/pipeline.js

```diff
@@ -26,6 +31,28 @@ function destroyer(stream, reading, writing, callback) {
 
   if (eos === undefined) eos = require('internal/streams/end-of-stream');
   eos(stream, { readable: reading, writable: writing }, (err) => {
+    // Check if this is a readable premature close.
+    if (
+      err &&
+      err.code === 'ERR_STREAM_PREMATURE_CLOSE' &&
+      reading &&
+      (stream._readableState && stream._readableState.ended) &&
+      (!writing || writableFinished(stream))
+    ) {
+      // Some readable streams will emit 'close' before 'end'. However, since
+      // this is on the readable side 'end' should still be emitted if the
+      // stream has been ended and no error emitted. This should be allowed in
+      // favor of backwards compatibility. Since the stream is piped to a
+      // destination this should not result in any observable difference.
+      stream
+        .on('end', () => {
+          closed = true;
+          callback();
+        })
+        .on('error', callback);
+      return;
+    }
```

Comment:
I would prefer if this logic lived in `end-of-stream` instead.

Reply:
Why? I can't think of a way to do it in `end-of-stream` while both keeping compat and "correct" behaviour at the same time. The current proposal I believe is achieving both.

Reply:
Ok

Reply:
@mafintosh: See here

Reply:
Not sure I completely follow this still